DISTS: Updated Webassembly / Emscripten Port

chkuendig opened the pull request:
I kept working on the Webassembly / Emscripten port in the past months and I feel it's ready now for another PR:

Changes include a few critical improvements:
- Dynamic Plugins: It's now possible to dynamically load plugins, making it possible to bundle all engines without the binary growing into unreasonable sizes. This depends on emscripten-core/emscripten#15893 which hasn't been merged, so it's manually patched in during the setup of the build environment.
- General Build Improvements: The build script is now fully self-contained and can be run with Github actions, this includes automatic download of games/demos as well as generating the default scummvm.ini for a demo page. See chkuendig/scummvm-demo on how a ScummVM demo app can be built (incl. playable demo).
- ScummvmFS: I added a custom filesystem layer which in the future can be extended to support range requests and persisting downloaded data for offline usage. (also, [jvilk/BrowserFS](https://github.com/jvilk/BrowserFS/) is not maintained anymore)

Minor Improvements:
- CLI arguments: It's now possible to pass CLI arguments as a anchor link. (this is used to generate the initial scummvm.ini file by opening `http://localhost:8080/scummvm.html#--add --path=/games --recursive` in a headless browser)
- HiDPI fix: SDL_GetDisplayDPI has been added to SDL for emscripten with libsdl-org/SDL#5383 - but this hasn't been released yet so I switched to using an unreleased version of SDL during the setup of the build environment.
- Gamepad API: There was a minor bug when opening the app on a non-secure context (i.e. HTTP instead of HTTPS) due to SDL wanting to use the Gamepad Web API, there's now a redirect to avoid this.
- Parts of the closed PR scummvm/scummvm#3686 are added to the SDL backend for fullscreen support, no exit buttons and proper URL opening.
- Director Engine: Yield back to the browser in the mail loop

I also added myself to the credits. Let me know if there's anything else required for me to officially take the maintainer role for this port.

As before, the latest version is running at http://scummvm.kuendig.io/ (now automatically built with Github Actions, see [chkuendig/scummvm-demo/](https://github.com/chkuendig/scummvm-demo/)).

chkuendig commented on backends/platform/sdl/posix/posix.cpp:
```diff
@@ -27,6 +27,9 @@
 #define FORBIDDEN_SYMBOL_EXCEPTION_system
 #define FORBIDDEN_SYMBOL_EXCEPTION_random
 #define FORBIDDEN_SYMBOL_EXCEPTION_srandom
+#ifdef __EMSCRIPTEN__
```
I might have misunderstood something, but the consensus in #3686 seemed to have been to not add another backend class. So I'm not sure what is preferred.

ccawley2011 replied:
The issue with that PR is that it that it duplicates a lot of stuff from the POSIX backend that looks like it wouldn't be useful on Emscripten (namely the path handling, the Linux AudioCD code and the use of fork/exec for opening the log file). Since having multiple subclasses of `OSystem_SDL` is intended for this exact purpose, I'd recommend one of the following approaches:
- If the POSIX-specific code is useful for Emscripten, then the Emscripten backend can inherit from `OSystem_POSIX` and override specific functions where needed.
- If the POSIX-specific code doesn't work or is overridden by the JavaScript code, then the Emscripten back end can inherit directly from `OSystem_SDL`, with simplified versions of `displayLogFile`, `addSysArchivesToSearchSet`, `getScreenshotsPath`, `getDefaultConfigFileName` and `getDefaultLogFileName` where necessary.

bluegr replied:
Totally agree with @ccawley2011's suggestions

chkuendig replied:
Thank you, I'll subclass `OSystem_POSIX` as most POSIX stuff can be reused - Emscripten is generally considered to be POSIX (and says so on [emscripten.org](https://emscripten.org)).

chkuendig replied:
I pushed a change implementing a new backend. Let me know what you think.

ccawley2011 commented on AUTHORS:
```diff
@@ -623,6 +623,9 @@ ScummVM Team
     Tizen / BADA:
        Chris Warren-Smith
 
+    Webassembly / Emscripten:
```
Updating the credits is normally done by editing `devtools/credits.pl`, followed by running `make credits` to update the other files.

chkuendig replied:
I updated `devtools/credits.pl` as well and ran `make credits`. Thank you for the hint!

chkuendig commented on dists/emscripten/build.sh:
```diff
@@ -1,206 +1,347 @@
 
   # HACK: the preload flags break emcc during configure as emcc enables NODERAWFS when run as part of configure
   # which doesn't support preloading assets, so we have to manually add those after configure to the config.mk file
+  LDFLAGS_LINKER=" --pre-js ./dists/emscripten/custom_shell-pre.js --post-js ./dists/emscripten/custom_shell-post.js --shell-file ./dists/emscripten/custom_shell.html "
```
This is what I've been looking for! It's now done this way with the latest commit.

chkuendig commented on configure:
```diff
@@ -5367,7 +5395,13 @@ if test "$_opengl_mode" != none ; then
+							_opengl_mode=gles2
+							_opengl_game_es2=yes
+							_opengl_glad=yes
+							_tinygl=no
```
I removed disabling TinyGL by default. I wanted to save the extra binary size since software rendering is quite slow, but I can easily do that by just adding `--disable-tinygl` to the demo build script

chkuendig commented on configure:
```diff
@@ -5367,7 +5395,13 @@ if test "$_opengl_mode" != none ; then
-
+						emscripten)
+							_opengl_mode=gles2
+							_opengl_game_es2=yes
```
Not sure about this one. This is required to enable gles2 by default afaik.

ccawley2011 replied:
Following commit https://github.com/scummvm/scummvm/commit/a9418f8f6343151ff3bf98df74a4b44c714c1b3e, `_opengl_game_es2` was removed and `_opengl_mode` was changed to set the OpenGL mode for both 2D and 3D games.

chkuendig replied:
Thank you, this is fixed with the newest push

chkuendig commented:
I incorporated the first set of feedback. 

FYI: I also noticed an issue with TLJ crashing when moving beyond the in game menu into the game - this came down to some weirdness with `glTexParameterf` which I couldn't explained. I switched to `glTexParameteri` as that a) fixed the issue, b) was the only place in all of ScummVM where `glTexParameterf` was used and c) is anyway the appropriate function since the parameters are integer enums (based on https://stackoverflow.com/questions/29919386/what-is-the-difference-between-gltexparameterf-and-gltexparameteri)

chkuendig commented:
I force-pushed a new set of (cleaned up) commits which should incorporate all feedback received so far and a few other improvements I made along the way.

lephilousophe reviewed (commented):
Some comments after skimming this PR

chkuendig commented on configure:
```diff
@@ -4498,7 +4529,7 @@ define_in_config_if_yes "$_flac" 'USE_FLAC'
 
 # Add the link to ogg only after vorbis, tremor and flac as it might be used by those.
-if test "$_ogg" = yes ; then
+if test "$_ogg" = yes && test "$_host_os" != "emscripten" ; then
```
Thank you for the feedback. I now changed it accordingly. The linker flags are removed if the respective [emscripten port](https://emscripten.org/docs/compiling/Building-Projects.html#emscripten-ports) is active.

lephilousophe replied:
Thanks! Looks good to me.

chkuendig commented:
FYI: I pushed a new revision of this PR. There's a few more things I might look into, but these things can wait.

Once I've tested the demo a bit more to discover any issue, I'd suggest this can be merged (I'll remove the draft flag accordingly)

aquadran commented on dists/emscripten/emscripten-15893-fix.patch:
```diff
@@ -0,0 +1,44 @@
+From 2ea11b4734de5656aaf3285145831e2bb9ad53ae Mon Sep 17 00:00:00 2001
```
scummvm repo is not good place to keep distro patches

chkuendig replied:
I agree this is a hack. Is it general not a good idea to include patches (and instead rely on only stable versions), or is this particular one a blocker since it isn't even in the PR of the upstream project? 

If it's the second, I would try to get this added to PR emscripten-core/emscripten#15893 (or start a second PR), so all patches have a corresponding PR. If we in general don't want to rely on unreleased code, this will mean:

- Either keep this PR open until all changes are merged and released in upstream
- Accept that certain things wont work with the merged code without manually applying these patches (this can be documented and a second PR could be put in draft until the changes all have landed)

The reason that these patches are checked into this repo, is because PRs aren't 100% stable (they are often rebased against master, which means they might not work against the stable release this script is depending on)

aquadran replied:
you can create upstream forks with patches applied and point to them. or put patches somewhere on GitHub and point to them

chkuendig replied:
I cleaned up the patches in 037536e8440e7931cdf62d1a0bf265e73ea644c1. All patches are now downloaded from Github (I still checked them in as well as PRs aren't 100% stable which could break the build).

aquadran replied:
I still see patches in PR

chkuendig replied:
I wasn't aware that patches in general aren't allowed. I removed them and uploaded them to another repo (to ensure they stay stable and don't change should the PRs change)

aquadran reviewed (changes requested):
scummvm repository is not good place to keep patches

aquadran reviewed (changes requested):
there are still patches in PR

sev- commented:
Looks great and I am backporting it to 2.6.0. Will you be able to build it at the time of release?

chkuendig commented:
Yes, this is easily built. I have Github Actions actually. I wonder through what artifacts we would provide? Im currently hosting on Github Pages and cant get more demos onto that (already hitting some limits), OTOH providing a zip file to download doesnt make much sense either.

neuromancer commented:
I made some quick PR to have this deployed: #3985. It will require some adjustments, but hopefully we have it merged to allow users to test 2.6.0 release candidate. Happy to include any improvements or suggestions from you!
